ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,215 @@
+
+  private volatile boolean closed = false;
+
+  public QueryMonitor(final KsqlConfig ksqlConfig, final KsqlEngine ksqlEngine) {
+    this(
+        ksqlConfig,
+        ksqlEngine,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, "QueryMonitor")),
+        CURRENT_TIME_MILLIS_TICKER
+    );
+  }
+
+  @VisibleForTesting
+  QueryMonitor(
+      final KsqlConfig ksqlConfig,
+      final KsqlEngine ksqlEngine,
+      final ExecutorService executor,
+      final Ticker ticker
+  ) {
+    this.retryBackoffMaxMs = ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS);
+    this.ksqlEngine = ksqlEngine;
+    this.executor = executor;
+    this.ticker = ticker;
+  }
+
+  public void start() {
+    executor.execute(new Runner());
+    executor.shutdown();
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+
+    try {
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  boolean isClosed() {
+    return closed;
+  }
+
+  void restartFailedQueries() {
+    ksqlEngine.getPersistentQueries().stream()
+        .forEach(query -> {
+          final QueryId queryId = query.getQueryId();
+          final KafkaStreams.State queryState = query.getState();
+
+          // If the query was restarted previously, check if it needs another restart; or if the
+          // query is now up and running, then remove it from the restart list.
+          if (queriesRetries.containsKey(queryId)) {
+            final RetryEvent retryEvent = queriesRetries.get(queryId);
+            if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+              if (queryState == KafkaStreams.State.ERROR) {
+                // Retry again if it's still in ERROR state
+                retryEvent.restart();
+              } else {
+                // Query is not in ERROR state anymore.
+                queriesRetries.remove(queryId);
+              }
+            }
+          } else if (queryState == KafkaStreams.State.ERROR) {
+            // Restart new query in ERROR state, and add it to the list of retries.
+            final RetryEvent retryEvent = new RetryEvent(
+                ksqlEngine, queryId, BASE_WAITING_TIME_MS, retryBackoffMaxMs, ticker);
+
+            queriesRetries.put(queryId, retryEvent);
+            retryEvent.restart();
+          }
+        });
+  }
+
+  private class Runner implements Runnable {
+    @Override
+    public void run() {
+      LOG.info("KSQL query monitor started.");
+
+      while (!closed) {
+        try {
+          restartFailedQueries();
+        } catch (final Exception e) {
+          LOG.warn("KSQL query monitor found an error attempting to restart failed queries.", e);
+        }
+      }
+    }
+  }
+
+  static class RetryEvent {
+    private final KsqlEngine ksqlEngine;
+    private final QueryId queryId;
+    private final Ticker ticker;
+
+    private int numRetries = 0;
```

Comment:
can be done in a different PR, but please create a ticket to make sure that we track these in our metrics - it would be good to see - and I think it should also be something that we can expose to the users so they know that they're in a crash-back-off loop